morphic.js

```diff
@@ -1052,6 +1052,185 @@
     - Jens Mönig
 */
 
+// Retina Canvas utilities /////////////////////////////////////////////
+
+// This installs a series of utilities that allow using Canvas the same way on
+// retina and non-retina displays. If the display is a retina one, the
+// underlying dimensions of the Canvas elements are doubled, but this will be
+// transparent to the code that uses Canvas. All dimensions read or written to
+// the Canvas element will be scaled appropriately.
+//
+// NOTE: This implementation is not exhaustive; it only implements what is
+// needed by the Snap! UI.
+
+(function() {
+    // Get the window's pixel ratio for canvas elements.
+    // See: http://www.html5rocks.com/en/tutorials/canvas/hidpi/
+    var ctx = document.createElement("canvas").getContext("2d"),
+            backingStorePixelRatio = ctx.webkitBackingStorePixelRatio ||
+                ctx.mozBackingStorePixelRatio ||
+                ctx.msBackingStorePixelRatio ||
+                ctx.oBackingStorePixelRatio ||
+                ctx.backingStorePixelRatio || 1;
+
+    // Unfortunately, it's really hard to make this work well when changing zoom level, so let's
+    // leave it like this right now, and stick to whatever the ratio was in the beginning.
+    var originalDevicePixelRatio = window.devicePixelRatio;
+
+    function getPixelRatio(imageSource) {
```

Comment:
pixel ratio will change if you drag the browser window from retina display to an external non-retina display.